src/main/java/net/minecraftforge/common/util/JsonMerger.java

```diff
@@ -0,0 +1,468 @@
+        {
+            if (!second.isJsonObject())
+            {
+                throw new JsonSyntaxException("The combining counterpart for a json object must be another json object");
+            }
+
+            if (mode == MergeMode.COMBINE)
+            {
+                return combineObjects(first.getAsJsonObject(), second.getAsJsonObject());
+            }
+
+            throw new JsonSyntaxException(String.format("Invalid combine mode for a json object: %s. Allowed: overwrite, combine", mode));
+        }
+        else if (first.isJsonArray())
+        {
+            switch(mode)
+            {
+                case COMBINE:
+                    return concatenateArrays(first.getAsJsonArray(), second.getAsJsonArray());
+                case ZIP:
+                    return zipArrays(first.getAsJsonArray(), second.getAsJsonArray());
+                default:
+                    throw new JsonSyntaxException(String.format("Invalid combine mode for a json array: %s. Allowed: overwrite, combine, zip", mode));
+            }
+        }
+        else if (first.isJsonNull())
+        {
+            if (mode != MergeMode.COMBINE)
+                throw new JsonSyntaxException(String.format("Invalid combine mode for a json null: %s", mode));
+        }
+        else
+        {
+            if (mode != MergeMode.COMBINE)
+                throw new JsonSyntaxException(String.format("Invalid combine mode for a json primitive: %s", mode));
+        }
+
+        return second;
+    }
+
+    public static JsonElement combineObjects(JsonObject first, JsonObject second)
+    {
+        JsonObject result = new JsonObject();
+        for(Map.Entry<String, JsonElement> entry : first.entrySet())
+        {
+            String key = entry.getKey();
+            JsonElement value = entry.getValue();
+            if (second.has(key))
+            {
+                result.add(key, combineJsonElements(value, second.get(key), MergeMode.COMBINE));
+            }
+            else
+            {
+                result.add(key, value);
+            }
+        }
+        for(Map.Entry<String, JsonElement> entry : first.entrySet())
+        {
+            String key = entry.getKey();
+            JsonElement value = entry.getValue();
+            if (!result.has(key))
+            {
+                result.add(key, value);
+            }
+        }
+        return result;
+    }
+
+    private static JsonElement zipArrays(JsonArray first, JsonArray second)
+    {
+        JsonArray a = new JsonArray();
+        int end = Math.min(first.size(),second.size());
+        for(int i=0;i<end;i++)
+        {
+            a.add(combineJsonElements(first.get(i), second.get(i), MergeMode.COMBINE));
+        }
+        return null;
+    }
+
+    private static JsonElement concatenateArrays(JsonArray first, JsonArray second)
+    {
+        List<JsonElement> elements = Lists.newArrayList();
+        for(int i=0;i<first.size();i++)
+            elements.add(first.get(i));
```

Comment:
Would it be worth it to add a check to these elements to ensure none of them accidentally contain `_forge_combine`? It shouldn't happen since I assume this will always be a vanilla file, but if used between two mods or something it could have some benefits.

Reply:
I used `"_forge_combine"` specifically to not have to worry about vanilla content using that key. While I could filter out those keys, I'm not sure it's worth the trouble.